docs/kuneiform/procedures.md

```diff
@@ -0,0 +1,275 @@
+    // do something
+    // ...
+    INSERT INTO users (id, name, age)
+    VALUES ($id, $name, $age);
+}
+```
+
+### Handling Arrays
+
+Procedures support basic array operations. It supports the following functions:
+
+- `array_append(type[], type) type[]`: appends a value to the end of an array.
+- `array_prepend(type, type[]) type[]`: prepends a value to the beginning of an array.
+- `array_cat(type[], type[]) type[]`: concatenates two arrays together.
+- `array_length(type[]) int`: returns the length of the array.
+
+**Arrays are 1-indexed.**
+
+```ts
+procedure handle_array() public view returns (val text) {
+    $arr text[] := ['hello', 'world'];
+    $arr := array_append($arr, 'goodbye');
+
+    $len int := array_length($arr); // $len == 3
+
+    // return the first value.
+    return $arr[1];
+}
+```
+
+### If/Then
+
+```ts
+procedure age_conditional($age int) public {
+    if $age < 0 {
+        // do something
+    } elseif $age < 18 {
+        // do something
+    } else {
+        // something else
+    }
+}
+```
+
+### For Loop
+
+For loops can be declared over 4 different targets: arrays, ranges, procedure calls, and SQL statements.
+
+#### Array Loop
+
+```ts
+procedure iterate_over_array($names text[]) public {
+    for $name in $names {
+        // $name is of type `text`
+    }
+}
+```
+
+#### Range Loop
+
+```ts
+procedure iterate_over_range() public {
+    // iterating over all numbers from 1 to 25
+    for $i in 1..25 {
+        // $i is type int
+    }
+}
+
+procedure iterate_range_dynamic($start int, $end int) public {
+    // can also iterate over dynamic ranges
+    for $i in $start..$end {
+        // ...
+    }
+}
+```
+
+#### Procedure Loops
+
+You can iterate over any SQL statement as well. The following examples reference an example table `users`, with columns `id`, `name`, and `age`.
+
+```ts
+procedure iterate_over_sql() public {
+    for $row in SELECT id, name, age from users {
+        // can access $row.id, $row.name, and $row.age
+    }
+}
+```
+
+You can also iterate of returning clauses:
+
+```ts
+procedure iterate_over_delete() public {
+    $age int := 100;
+    for $row in delete from users
+        where age = $age
+        returning * {
+        // can access $row.id, $row.name, and $row.age
+        // for all deleted users
+    }
+}
+```
+
+Procedures that return tables can also be iterated over:
+
+```ts
+procedure get_users() public view returns table(id uuid, name text, age int) {
+    // ...
+}
+
+procedure iterate_over_users() public {
+    for $row in SELECT * FROM get_users() {
+        // can access $row.id, $row.name, and $row.age
+    }
+}
+```
+
+#### Breaking From a Loop
+
+To conditionally exit a loop, you can use `break`:
+
+```ts
+procedure get_users() public {
+    $count int := 0;
+    for $row in SELECT * FROM users {
+        // do something
+
+        // break if too many records have been read
+        $count := $count + 1;
+        if $count >= 50 {
+            break
```

Comment:
I assume `break` works like any other sensible language with nested loops by breaking only the inner loop?  Also assuming nested loops are ok?

Do we have a `continue` do programmers just code around that with else and more indents?  That's fine, just curious.

Reply:
Nested breaks do work like any other language, and yes we do support nested loops. We do not support continue, however it would be easy to add.